### PR TITLE
Simplify RNN data preprocessing 

### PR DESCRIPTION
**Pull request by Steffy-zxf**
### PR types

Others 

### PR changes

Example

### Description

* build vocab from dataset instead to download the external vocab

### CODE REVIEW

File: examples/text_classification/rnn/README.md [hunk 1]
```diff
@@ -129,21 +129,21 @@ train_ds, dev_ds, test_ds = load_dataset("chnsenticorp", splits=["train", "dev",
 
 ### 模型训练
 
-在模型训练之前，需要先下载词汇表文件word_dict.txt，用于构造词-id映射关系。
+在模型训练之前，需要先下载停用词表文件word_dict.txt，用于构建词表过滤停用词。
+
+停用词表文件参考：
+https://github.com/YueYongDev/stopwords/blob/master/%E7%99%BE%E5%BA%A6%E5%81%9C%E7%94%A8%E8%AF%8D%E8%A1%A8.txt
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/README.md]: stopwords文件很小，可以考虑直接作为代码文件的一部分stopwords.txt 
然后用一个stopwords.py的python文件把这个文件加载进来，或者直接嵌入到python文件中。

File: examples/text_classification/rnn/train.py [hunk 2]
```diff
@@ -84,17 +85,24 @@ def create_dataloader(dataset,
...
     # Loads dataset.
     train_ds, dev_ds = load_dataset("chnsenticorp", splits=["train", "dev"])
+    texts = []
+    for data in train_ds:
+        texts.append(data["text"])
+    for data in dev_ds:
+        texts.append(data["text"])
+
+    # Reads stop words.
+    stop_words = read_stop_words(args.stop_words_path)
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/train.py]: 可以考虑将stopwords内容砍少一点，然后嵌入到stopwords.py文件中进行引用。

File: examples/text_classification/rnn/utils.py [hunk 3]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
         ids = tokenizer.encode(text)
         examples.append([ids, len(ids)])
     return examples
+
+
+def read_stop_words(file_path):
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: stopwords是个完整的英文单词，不是分开的两个单词

File: examples/text_classification/rnn/utils.py [hunk 4]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
...
+        stop_words = []
+        for line in f:
+            word = line.strip()
+            stop_words.append(word)
+    return stop_words
+
+
+def build_vocab(texts,
+                stop_words=[],
+                min_count=10,
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 建议使用min_freq

Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 新增一个参数num_words=None,  用于控制词表最大数量。

Comment (ZeyuChen) [reply examples/text_classification/rnn/utils.py]: 默认按照词频从大到小排序

File: examples/text_classification/rnn/utils.py [hunk 5]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
...
+        word_index (obj:`Dict`): The vocabulary from the corpus data.
+
+    """
+    documents = []
+    word_counts = defaultdict(int)
+    for text in texts:
+        if not text:
+            continue
+        doc = []
+        for word in jieba.cut(text):
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 改用Taskflow的word_segmentation来切词

File: examples/text_classification/rnn/utils.py [hunk 6]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
...
+            word_counts[word] += 1
+        documents.append(doc)
+
+    wcounts = []
+    for word, count in word_counts.items():
+        if count < min_count:
+            continue
+        wcounts.append((word, count))
+    wcounts.sort(key=lambda x: x[1], reverse=True)
+    sorted_voc = [pad_token, unk_token]
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py:112]: 这些特殊环节补充注释

File: examples/text_classification/rnn/utils.py (hunk 3 above)
Comment (ZeyuChen) [reply examples/text_classification/rnn/utils.py]: 可以考虑此处将stopwords作为一个基本的python set文件

File: examples/text_classification/rnn/utils.py [hunk 7]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
         ids = tokenizer.encode(text)
         examples.append([ids, len(ids)])
     return examples
+
+
+def read_stop_words(file_path):
+    with open(file_path, "r") as f:
+        stop_words = []
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 用list来存储，然后下面用in来判断是否在stopwords中，是一种低效的代码实现。

File: examples/text_classification/rnn/utils.py [hunk 8]
```diff
@@ -59,3 +63,55 @@ def preprocess_prediction_data(data, tokenizer):
...
+
+    """
+    documents = []
+    word_counts = defaultdict(int)
+    for text in texts:
+        if not text:
+            continue
+        doc = []
+        for word in jieba.cut(text):
+            if word in stop_words:
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: stopwords是个list，这样判断效率很低

File: examples/text_classification/rnn/utils.py (hunk 5 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: 使用jieba切词，建词表整体需要5.8s。
使用Taskflow的word_segmentation切词，建词表整体需要214.3s。

File: examples/text_classification/rnn/utils.py (hunk 4 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: 目前就是按照词频从大到小排序。

File: examples/text_classification/rnn/utils.py (hunk 7 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: 已经改用set存储。

File: examples/text_classification/rnn/utils.py (hunk 8 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: stopwords已经改用set存储

File: examples/text_classification/rnn/train.py (hunk 2 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/train.py]: Done

File: examples/text_classification/rnn/utils.py (hunk 3 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: Done

File: examples/text_classification/rnn/utils.py (hunk 4 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: Done

File: examples/text_classification/rnn/utils.py (hunk 6 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py:112]: Done

File: examples/text_classification/rnn/README.md [hunk 9]
```diff
@@ -180,7 +171,7 @@ python train.py --vocab_path='./senta_word_dict.txt' \
 
 以上参数表示：
 
-* `vocab_path`: 词汇表文件路径。
+* `vocab_path`: 用于保存根据预料库构建的词汇表的文件路径。
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/README.md]: 语料库

Comment (ZeyuChen) [inline examples/text_classification/rnn/README.md]: 语料库

File: examples/text_classification/rnn/README.md [hunk 10]
```diff
@@ -265,7 +258,7 @@ python predict.py --vocab_path='./senta_word_dict.txt' \
 如
 
 ```text
-Data: 这个宾馆比较陈旧了，特价的房间也很一般。总体来说一般      Lable: negative
+Data: 非常不错，服务很好，位于市中心区，交通方便，不过价格也高！      Lable: negative
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/README.md]: Label 不是 Lable， 看是否有全局的拼写错误整体修复下

Comment (ZeyuChen) [inline examples/text_classification/rnn/README.md]: Label 不是 Lable， 看是否有全局的拼写错误整体修复下

File: examples/text_classification/rnn/train.py [hunk 11]
```diff
@@ -23,16 +23,16 @@
...
 # yapf: disable
 parser = argparse.ArgumentParser(__doc__)
-parser.add_argument("--epochs", type=int, default=10, help="Number of epoches for training.")
+parser.add_argument("--epochs", type=int, default=15, help="Number of epoches for training.")
 parser.add_argument('--device', choices=['cpu', 'gpu', 'xpu', 'npu'], default="gpu", help="Select which device to train model, defaults to gpu.")
 parser.add_argument("--lr", type=float, default=5e-5, help="Learning rate used to train.")
 parser.add_argument("--save_dir", type=str, default='checkpoints/', help="Directory to save model checkpoint")
 parser.add_argument("--batch_size", type=int, default=64, help="Total examples' number of a batch for training.")
-parser.add_argument("--vocab_path", type=str, default="./senta_word_dict.txt", help="The directory to dataset.")
+parser.add_argument("--vocab_path", type=str, default="./vocab.json", help="The directory to vocab_path.")
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/train.py]: help那里，是directory吗？还是file path？

File: examples/text_classification/rnn/train.py [hunk 12]
```diff
@@ -84,17 +84,26 @@ def create_dataloader(dataset,
 
 if __name__ == "__main__":
     paddle.set_device(args.device)
-    set_seed()
+    set_seed(1000)
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/train.py:87]: 为什么需要set seed？

File: examples/text_classification/rnn/utils.py [hunk 13]
```diff
@@ -59,3 +66,53 @@ def preprocess_prediction_data(data, tokenizer):
...
+            filtered from the texts.
+        min_freq (obj:`int`): the minimum word frequency of words to be kept.
+        unk_token (obj:`str`): Special token for unknow token.
+        pad_token (obj:`str`): Special token for padding token.
+
+    Returns:
+        word_index (obj:`Dict`): The vocabulary from the corpus data.
+
+    """
+    documents = []
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 这个变量做什么用？

File: examples/text_classification/rnn/utils.py [hunk 14]
```diff
@@ -59,3 +66,53 @@ def preprocess_prediction_data(data, tokenizer):
...
+    word_counts = defaultdict(int)
+    for text in texts:
+        if not text:
+            continue
+        doc = []
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py]: 需要这个变量吗？

File: examples/text_classification/rnn/utils.py [hunk 15]
```diff
@@ -59,3 +66,53 @@ def preprocess_prediction_data(data, tokenizer):
...
+            word_counts[word] += 1
+        documents.append(doc)
+
+    wcounts = []
+    for word, count in word_counts.items():
+        if count < min_freq:
+            continue
+        wcounts.append((word, count))
+    wcounts.sort(key=lambda x: x[1], reverse=True)
+    if num_words is not None and len(wcounts) > (num_words - 2):
```
Comment (ZeyuChen) [inline examples/text_classification/rnn/utils.py:109]: num_words - 2 这段需要加注释

File: examples/text_classification/rnn/train.py (hunk 12 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/train.py:87]: 当时QA要求保证可稳定复现，所以固定随机种子。

File: examples/text_classification/rnn/utils.py (hunk 13 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: documents 本来用于记录text转word id后的文本表示。但考虑到后面需要配合`Pad`、`Stack`、`DataLoader`的用法（兼容以前版本，做最小改动），其实现在documents没有作用了。已删去。

File: examples/text_classification/rnn/utils.py (hunk 14 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py]: 已删除

File: examples/text_classification/rnn/README.md (hunk 1 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/README.md]: Done

File: examples/text_classification/rnn/README.md (hunk 9 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/README.md]: Done

File: examples/text_classification/rnn/README.md (hunk 10 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/README.md]: Done

Comment (Steffy-zxf) [reply examples/text_classification/rnn/README.md]: Done

File: examples/text_classification/rnn/train.py (hunk 11 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/train.py]: 已修改为file path

File: examples/text_classification/rnn/utils.py (hunk 15 above)
Comment (Steffy-zxf) [reply examples/text_classification/rnn/utils.py:109]: Done

**Review by ZeyuChen — APPROVED** [review]
LGTM
